ugs-core/src/com/willwinder/universalgcodesender/MarlinCommunicator.java

```diff
@@ -0,0 +1,90 @@
+package com.willwinder.universalgcodesender;
+
+import static com.willwinder.universalgcodesender.AbstractCommunicator.SerialCommunicatorEvent.COMMAND_SENT;
+
+import java.util.StringTokenizer;
+import java.util.logging.Logger;
+
+import com.willwinder.universalgcodesender.types.GcodeCommand;
+
+public class MarlinCommunicator extends BufferedCommunicator {
+
+	// TODO: this should be configurable via the UI rather than being a constant
+	private static final int MAX_MARLIN_ACTIVE_COMMANDS = 10;
+
+	private static final Logger logger = Logger.getLogger(MarlinCommunicator.class.getName());
+
+	private boolean marlinBusy = false;
+
+	public boolean isMarlinBusy() {
+		return marlinBusy;
+	}
+
+	public void setMarlinBusy(boolean marlinBusy) {
+		this.marlinBusy = marlinBusy;
+	}
+
+	@Override
+	public boolean allowMoreCommands() {
+		// protect Marlin's buffers against overflow
+		if (activeCommandListSize() >= MAX_MARLIN_ACTIVE_COMMANDS) {
+			return false;
+		}
+
+		// don't send an M0 from the g code file until the active queue has drained
+		if (this.getNextCommand().getCommandString().equals("M0")) {
+			return (this.activeCommandListSize() == 0);
+		}
+
+		if (marlinBusy) {
+			return false;
+		}
+		return super.allowMoreCommands();
+	}
+
+	protected MarlinCommunicator() {
+	}
+
+	@Override
+	public int getBufferSize() {
+		return 200; // TODO: what should it be?
```

Comment:
Browsed through the marlin code and it looks like they use 128 as default:
https://github.com/MarlinFirmware/Marlin/blob/8e1ea6a2fa1b90a58b4257eec9fbc2923adda680/Marlin/src/inc/Conditionals_adv.h

In GRBL we used 123 as we needed some space for real time commands and line breaks.